core/queryparser/sparql/src/main/java/org/eclipse/rdf4j/query/parser/sparql/SPARQLUpdateDataBlockParser.java

```diff
@@ -70,7 +72,90 @@ public RDFFormat getRDFFormat() {
 
 	@Override
 	protected void parseGraph() throws RDFParseException, RDFHandlerException, IOException {
-		super.parseGraph();
+		int c = readCodePoint();
+		int c2 = peekCodePoint();
+		Resource contextOrSubject = null;
+		boolean foundContextOrSubject = false;
+		if (c == '[') {
+			skipWSC();
+			c2 = readCodePoint();
+			if (c2 == ']') {
+				contextOrSubject = createNode();
+				foundContextOrSubject = true;
+				skipWSC();
+			} else {
+				unread(c2);
+				unread(c);
+			}
+			c = readCodePoint();
+		} else if (c == '<' || TurtleUtil.isPrefixStartChar(c) || (c == ':' && c2 != '-') || (c == '_' && c2 == ':')) {
+			unread(c);
+
+			Value value = parseValue();
+
+			if (value instanceof Resource) {
+				contextOrSubject = (Resource) value;
+				foundContextOrSubject = true;
+			} else {
+				// NOTE: If a user parses Turtle using TriG, then the following
+				// could actually be "Illegal subject name", but it should still
+				// hold
+				reportFatalError("Illegal graph name: " + value);
+			}
+
+			skipWSC();
+			c = readCodePoint();
+		} else {
+			setContext(null);
+		}
+
+		if (c == '{') {
+			setContext(contextOrSubject);
+
+			c = skipWSC();
+
+			if (c != '}') {
+				parseTriples();
+
+				c = skipWSC();
+
+				while (c == '.') {
+					readCodePoint();
+
+					c = skipWSC();
+
+					if (c == '}') {
+						break;
+					}
+
+					parseTriples();
+
+					c = skipWSC();
+				}
+
+				verifyCharacterOrFail(c, "}");
+			}
+		} else {
+			setContext(null);
+
+			// Did not turn out to be a graph, so assign it to subject instead
+			// and
+			// parse from here to triples
+			if (foundContextOrSubject) {
+				subject = contextOrSubject;
+				unread(c);
+				parsePredicateObjectList();
+			}
+			// Or if we didn't recognise anything, just parse as Turtle
+			else {
+				unread(c);
+				parseTriples();
+			}
+		}
+
+		if (c == '.' || c == '}') {
```

Comment:
@jeenbroekstra, we found that the failing testcases in ShaclTest, are due to this condition on readCodePoint function. Hence, we are not sure if these changes are sufficient. Kindly suggest an alternative solution or you can push a fix. Thanks for the help!

Reply:
The problem is that you are looking at the wrong character. You haven't advanced `c` yet, so it's still the last processed char - which will never be equal to '.' or '}', so the end char is never read. This is what causes the failures in the SHACL tests.

To fix, add this line before your if-condition:

 		c = peekCodePoint();

That sets the value of `c` to the next char to be read, and then we decide to actually read if it is indeed a full stop or a closign bracket.